CodeFlareSDK_Design_Doc.md

```diff
@@ -0,0 +1,113 @@
+# CodeFlare SDK Design Doc
+
+## Context and Scope
+
+The primary purpose for the CodeFlare SDK is to provide a pythonic interaction layer between a user and the CodeFlare Stack (a set of services that enable advanced queuing, resource management and distributed compute on a kubernetes cluster).
+
+The reason that this SDK is needed is due to the fact that many of the benefits associated with the CodeFlare stack are aimed at making the work of data scientists simpler and more efficient. However, since all parts of the CodeFlare stack are separate Kubernetes services, there needs to be something that unifies the interactions between the user and these separate services. Furthermore, we do not expect the average user to be experienced working with kubernetes infrastructure, and want to provide them with a python native way of interacting with these services.
+
+The SDK should support any operation that a user would need to do in order to successfully submit machine learning training jobs to their kubernetes cluster.
+
+## Goals
+
+* Serve as an interaction layer between a data scientist and the CodeFlare Stack (MCAD, InstaScale, KubeRay)
+* Abstract away user’s infrastructure concerns; Specifically, dealing with Kubernetes resources, consoles, or CLI’s (kubectl).
+* Provide users the ability to programmatically request, monitor and stop the kubernetes resources associated with the CodeFlare stack.
+
+## Non-Goals
+
+* We do not want to re-implement any existing SDK’s or clients for Ray, MCAD, or any other service.
+
+## Architecture and Design
+
+The CodeFlare SDK is a python package that allows a user to programmatically define, create, monitor, and shutdown framework clusters (RayClusters for now) as well as define, submit, monitor and cancel distributed training jobs for machine learning models on an authenticated Kubernetes cluster that has the CodeFlare Stack installed.
+
+In order to achieve this we need the capacity to:
+
+* Generate valid AppWrapper yaml files based on user provided parameters
+* Get, list, watch, create, update, patch, and delete AppWrapper custom resources on a kubernetes cluster
+* Get, list, watch, create, update, patch, and delete RayCluster custom resources on a kubernetes cluster.
+* Expose a secure route to the Ray Dashboard endpoint.
+* Define, submit, monitor and cancel Jobs submitted via TorchX. TorchX jobs must support both Ray and MCAD-Kubernetes scheduler backends.
+* Provide means of authenticating to a Kubernetes cluster
+
+![](/assets/images/sdk-diagram.png)
+
+### Framework Clusters:
+
+In order to create these framework clusters, we will start with a template AppWrapper yaml file with reasonable defaults that will generate a valid RayCluster via MCAD.
```

Comment:
Should we include this template appwrapper yaml at the end of the doc? Or have it documented/linked somewhere?